sklearn/pipeline.py

```diff
@@ -516,6 +529,103 @@ def _pairwise(self):
+    is advantageous when fitting is time consuming.
+
+    The purpose of the pipeline is to assemble several steps that can be
+    cross-validated together while setting different parameters.
+    For this, it enables setting parameters of the various steps using their
+    names and the parameter name separated by a '__', as in the example below.
+    A step's estimator may be replaced entirely by setting the parameter
+    with its name to another estimator, or a transformer removed by setting
+    to None.
+
+    Read more in the :ref:`User Guide <cached_pipeline>`.
+
+    Parameters
+    ----------
+    steps : list
+        List of (name, transform) tuples (implementing fit/transform) that are
+        chained, in the order in which they are chained, with the last object
+        an estimator.
+
+    memory : Instance of joblib.Memory or string (optional)
+        Used to cache the output of the computation of the tree.
+        By default, no caching is done. If a string is given, it is the
+        path to the caching directory.
+
+    Attributes
+    ----------
+    named_steps : dict
+        Read-only attribute to access any step parameter by user given name.
+        Keys are step names and values are steps parameters.
+
+    Examples
+    --------
+    >>> from sklearn.datasets import samples_generator
+    >>> from sklearn.decomposition import PCA
+    >>> from sklearn.ensemble import RandomForestClassifier
+    >>> from sklearn.externals.joblib import Memory
+    >>> from sklearn.model_selection import GridSearchCV
+    >>> from sklearn.pipeline import CachedPipeline
+    >>> from tempfile import mkdtemp
+    >>> # setup a temporary directory
+    >>> cachedir = mkdtemp()
+    >>> # setup the cache
+    >>> memory = Memory(cachedir=cachedir, verbose=0)
+    >>> # generate some data to play with
+    >>> X, y = samples_generator.make_classification(
+    ...     n_samples=100, n_informative=5, n_redundant=0, random_state=42)
+    >>> # Create a pipeline with a PCA and an RandomForestClassifier
+    >>> pca = PCA()
+    >>> clf = RandomForestClassifier()
+    >>> pipeline = CachedPipeline([('pca', pca), ('rf', clf)],
+    ...                           memory=memory)
+    >>> # Make a grid_search of the best parameters
+    >>> parameters = {'pca__n_components': (.25, .5),
+    ...               'rf__n_estimators': (10, 20)}
+    >>> # To avoid refitting the same PCA for different configuration
+    >>> # CachedPipeline will store the PCA estimator and load it
+    >>> # when required durin the grid search
+    >>> grid_search = GridSearchCV(pipeline, parameters, verbose=0)
+    >>> grid_search.fit(X, y)
+    ...                                              # doctest: +ELLIPSIS
+    GridSearchCV(...)
+
+    See also
+    --------
+    Pipeline
+        Pipeline of transforms with a final estimator.
+
+    """
+
+    def __init__(self, steps, memory=Memory(cachedir=None, verbose=0)):
```

Comment:
I think we should not pass mutable objects as default value for a kwarg: if one instance of the `CachedPipeline` changes the value of `memory.cachedir`, all the other instances of `CachedPipeline` (with the default value) will be impacted for instance.

http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments